go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -502,6 +502,11 @@ func (s *VtctldServer) backupTablet(ctx context.Context, tablet *topodatapb.Tabl
 				return err
 			}
 
+			// Check again that we haven't become primary during the backup
+			if tabletInfo.Type == topodatapb.TabletType_PRIMARY {
+				return nil
+			}
+
```

Comment:
I believe the right fix is to actually not call `SetReplicationSource` here at all. The `ChangeTabletType` code path already effectively does that here
https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletmanager/rpc_actions.go#L100

Reply:
Though.. I'm guessing that this line is present to handle the case where the identity of the primary changed while the replica was taking a backup.
Even then, it doesn't belong at this outer layer, it should all be handled inside the call to `tmc.Backup` because that is where we are making the decisions of whether the tablet type needs to change or not, whether we stop MySQL or not etc.

Reply:
Also, it will be good if you can create a unit test to validate the final change. Either in https://github.com/vitessio/vitess/blob/main/go/vt/vtctl/grpcvtctldserver/server_test.go#L510 or in the `tabletmanager` package.

Reply:
Hello Olga!,
I looked at the PR today and realized that the bug you are running into came to be because of a PR I did. https://github.com/vitessio/vitess/pull/9725 was the PR that added the `SetReplicationSource` RPC at the end of the Backup call. The motivation for the change was to fix the semi-sync settings on the vttablets after the Backup process has ended. If a vttablet becomes a REPLICA type from BACKUP, it might need to start sending semi-sync ACKs. 
The addition of this however broke backups from xtrabackup since the tablet itself can become a PRIMARY which the code doesn't take into account.

As @deepthi said, it doesn't look like a good idea to keep adding code here in the vtctld layer. Instead, we should remove the `SetReplicationSource` call entirely from here and instead, fix the semi-sync and replication settings on the vttablet directly. 

I think the best place for these changes would be https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletmanager/rpc_backup.go#L101. Here we can read the shard record and keyspace record to find the current primary and make necessary actions.

Regarding testing, I think tests can be added in https://github.com/vitessio/vitess/blob/main/go/vt/wrangler/testlib/backup_test.go as the original PR does.